### PR TITLE
Implement worst case scenario for price algorithm v1

### PR DESCRIPTION
**Pull request by rafal-ch**
## Linked Issues/PRs
Closes https://github.com/FuelLabs/fuel-core/issues/2203

## Description
This PR adds the `worst_case()` calculation to the `V1` price algorithm.

## Checklist
- [X] New behavior is reflected in tests

### Before requesting review
- [X] I have reviewed the code myself

### CODE REVIEW

File: crates/fuel-gas-price-algorithm/src/lib.rs [hunk 1]
```diff
@@ -4,3 +4,24 @@
 
 pub mod v0;
 pub mod v1;
+
+#[allow(clippy::cast_possible_truncation)]
+pub(crate) fn cumulative_percentage(
```
Comment (MitchTurner) [inline crates/fuel-gas-price-algorithm/src/lib.rs:9]: I think we should name this "cumulative_percentage_change" or something like that instead

Comment (rafal-ch) [reply crates/fuel-gas-price-algorithm/src/lib.rs]: Updated in https://github.com/FuelLabs/fuel-core/pull/2219/commits/adffd995e4949ba2e025d88ea6c51660187e1d7d

**Review by MitchTurner — DISMISSED** [review]

File: crates/fuel-gas-price-algorithm/src/lib.rs [hunk 2]
```diff
@@ -4,3 +4,24 @@
...
+    let blocks = height.saturating_sub(for_height) as f64;
+    let percentage_as_decimal = percentage as f64 / 100.0;
+    let multiple = (1.0f64 + percentage_as_decimal).powf(blocks);
+    let mut approx = new_exec_price as f64 * multiple;
+    // account for rounding errors and take a slightly higher value
+    const ARB_CUTOFF: f64 = 16948547188989277.0;
+    if approx > ARB_CUTOFF {
+        const ARB_ADDITION: f64 = 2000.0;
+        approx += ARB_ADDITION;
+    }
```
Comment (Dentosal) [inline crates/fuel-gas-price-algorithm/src/lib.rs:24]: I don't understand what happens here. What's the meaning of `ARB_CUTOFF`?

Comment (MitchTurner) [reply crates/fuel-gas-price-algorithm/src/lib.rs:24]: It's a hack for dealing with really large floats. The rounding error will be on the order of `2000` and based on the prop tests will always result in a value that is larger than the `actual` worst case value. 

So `ARB_CUTOFF` is where we start seeing the error and `ARB_ADDITION` is the the amount we add to account for that error. This is the cost of using floats to represent these values, maybe there is a cleaner way to represent it :).

Comment (netrome) [reply crates/fuel-gas-price-algorithm/src/lib.rs:24]: It doesn't sound arbitrary then. Can we rename it to something like this?
```suggestion
    // account for rounding errors and take a slightly higher value
    const ROUNDING_ERROR_CUTOFF: f64 = 16948547188989277.0;
    if approx > ROUNDING_ERROR_CUTOFF {
        const ROUNDING_ERROR_COMPENSATION: f64 = 2000.0;
        approx += ROUNDING_ERROR_COMPENSATION;
    }
```

Comment (MitchTurner) [reply crates/fuel-gas-price-algorithm/src/lib.rs:24]: Background:

https://github.com/FuelLabs/fuel-core/pull/2025#discussion_r1681725245

File: crates/fuel-gas-price-algorithm/src/lib.rs [hunk 3]
```diff
@@ -4,3 +4,24 @@
 
 pub mod v0;
 pub mod v1;
+
+#[allow(clippy::cast_possible_truncation)]
+pub(crate) fn cumulative_percentage_change(
```
Comment (netrome) [inline crates/fuel-gas-price-algorithm/src/lib.rs:9]: It feels unclean to have this function on the crate root. How about putting it in a `utils` module or something similar?

File: crates/fuel-gas-price-algorithm/src/v1/tests/algorithm_v1_tests.rs [hunk 4]
```diff
@@ -28,3 +32,60 @@ fn calculate__returns_sum_of_da_and_exec_gas_prices() {
...
+
+    let expected = expected_exec_price.saturating_add(expected_da_gas_price);
+
+    dbg!(actual, expected);
+    assert!(actual >= expected);
+}
+
+proptest! {
+    #[test]
+    fn worst_case__correctly_calculates_value(
```
Comment (MitchTurner) [inline crates/fuel-gas-price-algorithm/src/v1/tests/algorithm_v1_tests.rs:81]: We should add another test to check for overflows. See the V0 test and this convo:
https://github.com/FuelLabs/fuel-core/pull/2025#discussion_r1683965151

**Review by netrome — DISMISSED** [review]
Some suggestions, otherwise LGTM

File: crates/fuel-gas-price-algorithm/src/lib.rs (hunk 2 above)
Comment (MitchTurner) [reply crates/fuel-gas-price-algorithm/src/lib.rs:24]: > It doesn't sound arbitrary then. Can we rename it to something like this?

There isn't a "correct" value to choose here, that's why I called it `ARB`, but you're right that the reason we chose it isn't completely arbitrary and your suggestion is much clearer and helpful anyway.

Comment (rafal-ch) [reply crates/fuel-gas-price-algorithm/src/lib.rs]: Variable names updated in https://github.com/FuelLabs/fuel-core/pull/2219/commits/4337b14502861275c52aabedc4d9396b6f9e5147

File: crates/fuel-gas-price-algorithm/src/lib.rs (hunk 3 above)
Comment (rafal-ch) [reply crates/fuel-gas-price-algorithm/src/lib.rs]: Updated in https://github.com/FuelLabs/fuel-core/pull/2219/commits/d1672d32199c4d29da6c2430bfef9f1d47937e8c

File: crates/fuel-gas-price-algorithm/src/v1/tests/algorithm_v1_tests.rs (hunk 4 above)
Comment (rafal-ch) [reply crates/fuel-gas-price-algorithm/src/v1/tests/algorithm_v1_tests.rs:81]: Test added in https://github.com/FuelLabs/fuel-core/pull/2219/commits/6cb31617facd5b55ed99b649146c2750654eb133

File: crates/fuel-gas-price-algorithm/src/v1/tests/algorithm_v1_tests.rs [hunk 5]
```diff
@@ -28,3 +32,114 @@ fn calculate__returns_sum_of_da_and_exec_gas_prices() {
...
+        let target_height = starting_height.saturating_add(block_horizon);
+        let _ = algorithm.worst_case(target_height);
+
+        // then
+        // doesn't panic with an overflow
+    }
+}
+
+#[test]
+fn worst_case__same_block_gives_new_exec_price() {
```
Comment (MitchTurner) [inline crates/fuel-gas-price-algorithm/src/v1/tests/algorithm_v1_tests.rs:122]: This name is wrong, it's misssing the da portion in the name. I think instead just saying it gives the same as calculate() would be best, for both the name and how it calculates the expected value.

Comment (rafal-ch) [reply crates/fuel-gas-price-algorithm/src/v1/tests/algorithm_v1_tests.rs]: Names updated in https://github.com/FuelLabs/fuel-core/pull/2219/commits/c55aa6569f1a001544c96562dd4fd16561b9be26

**Review by MitchTurner — APPROVED** [review]

**Review by netrome — APPROVED** [review]

**Pull request merged**